### PR TITLE
SOT-70: FRI

### PR DESCRIPTION
**Pull request by bjergsen243**
SOT-70: FRI

# Check List

- [x] Did you check your grammar / reader proof your document using ChatGPT?
- [x] Don't forget to squash commits into meaningful chunks before merging
- [x] After merging this, please update your task status!
- [x] Did you verify that your math formulas are rendered correctly?

### CODE REVIEW

File: docs/fri.md [hunk 1]
```diff
@@ -1 +1,137 @@
...
+
+Subset $\ohm$ has size $n = \rho^{-1}k$, comprises all $n$-th [roots of unity](../docs/plonk.md#Roots%20of%20Unity) in $\mathbb{F}_p$.
+
++ $k$ is the degree of $q$.
++ Assume $n$ is a power of 2.
++ $\rho^{-1} \geq 2$ is called the "FRI blowup factor".
++ $\rho \leq 1/2$ is called the "rate of the Reed-Solomon code" used.
++ $\lambda$ is the security parameter a.k.a "$\lambda$ bits of security".
+
+> [!NOTE] Tradeoffs in blowup factor
```
Comment (sonntuet1997) [inline docs/fri.md:58]: Fix this

File: docs/fri.md [hunk 2]
```diff
@@ -1 +1,137 @@
...
+> The bigger $\rho^{-1}$ is:
+> 
+> + The slower prover is
+> + The smaller verifier costs will be
+
+Let $\omega \in \mathbb{F}_p$ be a primitive $n$'th root of unity. That
+is, $n$ is the smallest integer such that $\omega^n = 1$. Then $\ohm = \{1, \omega, \omega^2, ..., \omega^{n-1} \}$.
+
+$\ohm$ is a "multiplicative subgroup" of $\mathbb{F}_q$
```
Comment (sonntuet1997) [inline docs/fri.md:67]: Add link refers to `multiplicative subgroup`

File: docs/fri.md [hunk 3]
```diff
@@ -1 +1,137 @@
...
+
++ If $x$ and $y$ are both $n$'th roots of unity, then so is $xy$.
++ Since $n$ is even: if $x$ is a $n$'th root of unity, $x^2$ is a ($n/2$)'th root of unity.
++ Since $n$ is even: if $x$ is a $n$'th root of unity, so is $-x$.
+
+$\ohm$ has size $n$ if and only if $n$ divides $p - 1$
```
Comment (sonntuet1997) [inline docs/fri.md:73]: Do we have the explanation for this ?

File: docs/fri.md [hunk 4]
```diff
@@ -1 +1,137 @@
...
+
++ $p - 1$ is divisible $2^{32}$.
```
Comment (sonntuet1997) [inline docs/fri.md:75]: $2^{32}$ ?

File: docs/fri.md [hunk 5]
```diff
@@ -1 +1,137 @@
...
++ Running FRI over the field can support any power-of-two value of $n$ up to $2^{32}$.
+
+> [!NOTE] Element in domain $\ohm$
+> To let FRI work, we need the property that for every $\omega$ in the domain $\ohm$, it holds that $-\omega$ is also in $\ohm$.
```
Comment (sonntuet1997) [inline docs/fri.md:79]: check grammar

File: docs/fri.md [hunk 6]
```diff
@@ -1 +1,137 @@
...
++ If $x$ and $y$ are both $n$'th roots of unity, then so is $xy$.
++ Since $n$ is even: if $x$ is a $n$'th root of unity, $x^2$ is a ($n/2$)'th root of unity.
++ Since $n$ is even: if $x$ is a $n$'th root of unity, so is $-x$.
+
+$\ohm$ has size $n$ if and only if $n$ divides $p - 1$
+
++ $p - 1$ is divisible $2^{32}$.
++ Running FRI over the field can support any power-of-two value of $n$ up to $2^{32}$.
+
+> [!NOTE] Element in domain $\ohm$
```
Comment (sonntuet1997) [inline docs/fri.md:78]: Fix this

File: docs/fri.md [hunk 7]
```diff
@@ -1 +1,137 @@
...
+### Folding Phase (or Commit Phase)
+
+#### High Level Idea
+
+Verifier picks a random field element $r$, and uses it to "randomly combine" every two paired-up entries.
+
+The prover then Merkle-commits to the folded vector. The random combining technique is
+chosen so that the folded vector will have half the degree of the original vector.
+
+This folding process is repeated util the degree should falls to 0.
```
Comment (sonntuet1997) [inline docs/fri.md:90]: Rewrite this sentence

File: docs/fri.md [hunk 8]
```diff
@@ -1 +1,137 @@
...
+
+At this point, the length of the folded vector is still $\rho^{-1} \geq 2$. However, since the degree should
+be 0, the prover can specify the folded vector with a single field element.
+
+The folding phase consists of $\log(k)$ rounds.
+
+#### Detail
+
+Split $q(X)$ into "even and odd" parts in the following sense
```
Comment (sonntuet1997) [inline docs/fri.md:99]:
```suggestion
Split $q(X)$ into "even and odd" parts in the following sense:
```

File: docs/fri.md [hunk 9]
```diff
@@ -1 +1,137 @@
...
+
++ $q(X) = q_e(X^2) + Xq_o(X^2)$
++ E.g., if $q(X) = 1 + 2x + 3X^2 + 4X^3$
```
Comment (sonntuet1997) [inline docs/fri.md:102]:
```suggestion
    + E.g., if $q(X) = 1 + 2x + 3X^2 + 4X^3$
```

File: docs/fri.md [hunk 10]
```diff
@@ -1 +1,137 @@
...
+  + Then $q_e(X) = 1 + 3X$ and $q_o(X) = 2 + 4X$
+  + Both $q_e$ and $q_o$ have (at most) half the degree of $q$
++ Verifier picks a random field element $r$ and sends $r$ to prover.
++ The prescribed "folding" $q$ is: $q_{fold}(Z) = q_e(Z) + rq_o(Z)$.
++ The degree of $q_{fold}$ is half the degree of $q$ itself.
+
+> [!NOTE] The domain of $q_{fold}$
```
Comment (sonntuet1997) [inline docs/fri.md:109]: Fix this

File: docs/fri.md [hunk 11]
```diff
@@ -1 +1,137 @@
...
+chosen so that the folded vector will have half the degree of the original vector.
+
+This folding process is repeated util the degree should falls to 0.
+
+At this point, the length of the folded vector is still $\rho^{-1} \geq 2$. However, since the degree should
+be 0, the prover can specify the folded vector with a single field element.
+
+The folding phase consists of $\log(k)$ rounds.
+
+#### Detail
```
Comment (sonntuet1997) [inline docs/fri.md:97]:
```suggestion
#### Details
```

File: docs/fri.md [hunk 12]
```diff
@@ -1 +1,137 @@
...
++ The prescribed "folding" $q$ is: $q_{fold}(Z) = q_e(Z) + rq_o(Z)$.
++ The degree of $q_{fold}$ is half the degree of $q$ itself.
+
+> [!NOTE] The domain of $q_{fold}$
+> The commitment on $q_{fold}$ will not be over $\ohm$ but over $\ohm^2 = {\omega^2: \omega \in \ohm}$.
+> 
+> $\ohm^2$ will also have to satisfy the $\{ \omega, -\omega \}$ property, and so on.
+
+Let $x$ and $-x$ be $n$'th roots of unity and $z = x^2$. Then:
+$$q_{fold}(z) = \frac{r + x}{2x}q(x) + \frac{r - x}{-2x}q(-x)$$
```
Comment (sonntuet1997) [inline docs/fri.md:115]: Link to the explanation of the formula of the RISC Zero article

**Review by sonntuet1997 — APPROVED** [review]

File: docs/fri.md [hunk 13]
```diff
@@ -1 +1,137 @@
...
+
+Prover Merkle-commits to evaluations $q(x)$ for those $x$ in a subset $\ohm$ of $\mathbb{F}_p$.
+
+Subset $\ohm$ has size $n = \rho^{-1}k$, comprises all $n$-th [roots of unity](../docs/plonk.md#Roots%20of%20Unity) in $\mathbb{F}_p$.
+
++ $k$ is the degree of $q$.
++ Assume $n$ is a power of 2.
++ $\rho^{-1} \geq 2$ is called the "FRI blowup factor".
++ $\rho \leq 1/2$ is called the "rate of the Reed-Solomon code" used.
++ $\lambda$ is the security parameter a.k.a "$\lambda$ bits of security".
```
Comment (ktokto313) [inline docs/fri.md:56]: We have the description of $\lambda$ here but i don't see the formula with $\lambda$ anywhere

**Review by ktokto313 — APPROVED** [review]
I couldn't understand polynomial commitment from FRI

File: docs/fri.md (hunk 13 above)
Comment (bjergsen243) [reply docs/fri.md:56]: I mentioned it in the query phase section

**Review by hduoc2003 — APPROVED** [review]

File: docs/fri.md [hunk 14]
```diff
@@ -1 +1,159 @@
...
++ [FRI Mechanics by RISC Zero](https://www.youtube.com/watch?v=wqRuoyH3Mqk&t=1032s&ab_channel=RISCZero)
++ [FRI explanation with example by RISC Zero](https://dev.risczero.com/proof-system/stark-by-hand#lesson-11-fri-protocol-commit-phase)
++ [FRI lecture by ZK Learning](https://www.youtube.com/watch?v=A3edAQDPnDY&ab_channel=Blockchain-Web3MOOCs)
++ [FRI Slide by ZK Learning](https://zk-learning.org/assets/lecture8.pdf)
+
+## Low Degree Testing
+
+Low degree testing refers to the problem of deciding whether a given function is a polynomial of some bounded degree.
+
+FRI protocol uses the idea of testing two polynomials with half the queries and splitting a polynomial into two smaller ones.
```
[the remark this answers is not in the document]
Comment (bjergsen243) [reply docs/fri.md:16]: you can read it here
https://medium.com/starkware/low-degree-testing-f7614f5172db

**Pull request merged**